oeps/oep-0026-arch-xapi-integration.rst

```diff
@@ -0,0 +1,483 @@
+
+In the short term, however, we will not implement our own LRS since the initial integration
+efforts do not require this (as shown in the first diagram above).
+
+Statement API
+=============
+
+xAPI specifies 4 standard REST-ful JSON payload APIs, of which only the `Statement API`_
+is needed for our purposes as that API is sufficient for tracking learning activity.
+
+The Statement API has essentially the following parts: `xAPI Actor`_, `xAPI Verb`_,
+`xAPI Object`_, `xAPI Context`_, and `xAPI Result`_.
+
+.. _Statement API: https://xapi.com/statements-101/
+
+xAPI Actor
+~~~~~~~~~~
+Although the **Actor** field can be either an *Agent* or a *Group*, we will primarily
+support only the *Agent* type, which is used for individuals performing an activity
+(`xAPI Verb`_ on an `xAPI Object`_).
+
+An Actor can be identified using `Friend of a Friend (FOAF)`_ vocabulary with either:
+(1) `email address`_, (2) `hash of email address`_, (3) `OpenID URI`_, or (4) `account`_
+with a *homepage*-scoped identifier.  One of these is sent along with the Actor's "name".
+To be mindful of learner privacy, we will initially take a conservative approach
+and only send #3, with an Open edX anonymized unique identifier of the learner
+(`Anonymized User ID`_).
+
+In the future, if certain external systems require `Personally Identifiable Information
+(PII)`_, like the learner's email address or name, then those may be conditionally sent
+with appropriate permissions. Adaptive engines, however, do not need PII.
+
+Initially, we will exclude the "name" field. However, if we find that xAPI JSON parsers
+assume this field always exists, then we can include the field but provide a non-PII
+value, such as a copy of the `Anonymized User ID`_.
+
+Anonymized User ID
+^^^^^^^^^^^^^^^^^^
+An *Anonymized User ID* will be generated as a SHA-256 hash of the Open edX user_id, 
+which is the **id** value of the user's row in the `Django auth_user`_ table in the
+Open edX LMS.
+
+As mentioned in `Using consistent user identifiers in Segment events`_, the other user
+identifiers in the Open edX platform aren't perfectly suited for inclusion in externally
+sent events:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Identifier
+     - Limitations
+   * - username
+     - Often contains PII, such as user's name or year of birth.
+   * - LMS user_id
+     - Information leakage of auth_user table with easily guessable values; tied down
+       to implementation of the housing database.
+   * - email address
+     - Is PII and modifiable by the user.
+   * - anonymous user_id
+     - It is currently constructed by hashing the user's LMS user_id with the Django
+       server's *SECRET_KEY* value. This value will change when the *SECRET_KEY* is rotated.
+
+.. _Django auth_user: https://docs.djangoproject.com/en/2.0/topics/auth/default/#user-objects
+
+Example
+^^^^^^^
+
+Here is an example of an **Actor** JSON value that we would generate:
```

Comment:
When there is a guest user (we have some open activities that can be reached by guess) we are sending "null" as the user id. let me know if it works for you - or if you want to have another solution,

Reply:
Interesting.  We don't have that situation ourselves today - although we do plan to support "anonymous access to courseware".  Though, wouldn't you want to distinguish events from one anonymous user from another anonymous user?